api-http/proposals/http-plugin.md

```diff
@@ -0,0 +1,186 @@
+using api_entry = std::pair<url_path_t, url_handler>;
+class http_plugin : public appbase::plugin<http_plugin>
+ {
+    public:
+
+      void add_handler(const string& url, const url_handler&, appbase::exec_queue q, int priority = appbase::priority::medium_low, http_content_type content_type = http_content_type::json);
+      void add_api(const api_description& api, appbase::exec_queue q, int priority = appbase::priority::medium_low, http_content_type content_type = http_content_type::json) {
+         for (const auto& call : api)
+            add_handler(call.first, call.second, q, priority, content_type);
+      }
+
+      void add_async_handler(const string& url, const url_handler& handler, http_content_type content_type = http_content_type::json);
+      void add_async_api(const api_description& api, http_content_type content_type = http_content_type::json) {
+         for (const auto& call : api)
+            add_async_handler(call.first, call.second, content_type);
+      }
+}
+```
+
+It will be converted as follows in this proposal
+```
+   using server_address_t = std::string;
+   using url_path_t = std::string;
+   using request_body_t = std::string;
+   using url_handler = std::function<void(const server_address_t&, url_path_t&&, request_body_t&&, url_response_callback&&)>;
+
+   enum class api_category {
+     node,
+     get_info,
+     chain_ro,
+     chain_rw,
+     net_ro,
+     net_rw,
+     producer_ro,
+     producer_rw,
+     snapshot,
+     trace_api,
+     prometheus,
+   };
+
+   struct api_entry {
+     api_category        category;
+     std::string         server_address; // empty value mean the API is available for all listened addresses, NOT DISABLE
+     std::string         path;
+     url_handler         handler;
+     appbase::exec_queue q;
+     int                 priority = appbase::priority::medium_low
+     http_content_type   content_type = http_content_type::json;
+   };
+
+   class http_plugin : public appbase::plugin<http_plugin> {
+     public:
+       void add_sync_api(api_entry&& api);
+       void add_async_api(api_entry&& api);
+   };
+```
+
+## server address format
+* server address can be in one of the following formats
+  - ipv4_address:port like 127.0.0.1:8080
+  - ipv6_address:port like [2001:db8:3c4d:15::1a2f:1a2b]:8080
+  - hostname:port like my.domain.com:8080
+  - unix socket path (must starts with '/' or './')
+  - http-server-address
+
+* In the case of the hostname, the IP address is resolved by ARP, *ALL* resolved addresses will be listened to.
+* Whether an IPv4-mapped IPv6 address can be used to handle both IPv4 connections and IPv6 connections is determined by system configuration. Some tools, like Apache server, have the option `-—enable-v4-mapped` to change the behavior, should we do the same?
+* To listen to all interfaces, just use 0.0.0.0:8008 or [::1]:8080; alternative syntax just adds unnecessary complexity.
```

Comment:
Not sure I agree with that, if I want to listen on all IPv6 & IPv4 why do I have to make two config entries vs a `*:8080`?